Browser default actions (#212)
javascript-tutorial/it.javascript.info: 2-ui/2-events/04-default-browser-action/article.md

# Azioni predefinite del browser

Molti eventi vengono ricondotti verso determinate azioni del browser.

Per esempio:

- Un click su un link - inizializza la navigazione verso il suo URL.

> - Un click su un pulsante di invio di un form - inizializza l'invio dello stesso al server.

Author: /done

> - Premendo e spostando il pulsante su un testo - lo si seleziona.

Author: mi prendo l'appunto e correggo.

> - Premendo e spostando il pulsante su un testo - lo seleziona.

Author: /done

> Quando gestiamo un evento con JavaScript, potremmo non volere affatto che la corrispondente azione del browser avvenga, ed implementare invece un altro comportamento.

Author: /done

## Prevenire le azioni del browser

> Ci sono due maniere per comunicare al browser che non vogliamo che esegua alcuna azione:

Author: /done

> - La maniera principale è quella di usare l'oggetto `event`, dentro il quale c'è il metodo `event.preventDefault()`.

Author: /done

</ul>
```

Ecco come appare applicando qualche stile CSS:

[iframe height=70 src="menu" link edit]

> Gli elementi del menù sono implementati come links HTML `<a>`, e non pulsanti `<button>`. Ci sono tante ragioni per fare ciò, per esempio:

Author: /done

- A molte persone piace usare "tasto destro" -- "apri in una nuova finestra". Se usassimo `<button>` oppure `<span>`, questa funzionalità non potrebbe essere usata.
- I motori di ricerca seguono i link `<a href="...">` nel processo di indicizzazione.

Questo è il motivo per cui usiamo `<a>` nel markup. Ma dato che normalmente intendiamo gestire i click tramite JavaScript, dovremo prevenire le azioni predefinite del browser.

Come in questo caso:

```js
menu.onclick = function(event) {
  if (event.target.nodeName != 'A') return;

  let href = event.target.getAttribute('href');
  alert( href ); // ...possono derivare dai dati scaricati dal server, dalla generazione della UI, etc. 

*!*
  return false; // previene l'azione del browser (non reindirizza verso l'URL)
*/!*
};
```

Se omettessimo `return false`, subito dopo l'esecuzione del nostro codice, il browser compirebbe la sua "azione predefinita" -- navigando quindi verso l'URL impostato nell'`href`. E non è ciò che vogliamo, dal momento che stiamo gestendo noi stessi il click.

> A proposito, qui l'utilizzo dell'event delegation rende il nostro menù molto flessibile. Possiamo aggiungere liste annidate e stilizzare usando i CSS per farle "scendere a mo' di tendina".

Author: /done

Perché dovrebbe essere necessaria una cosa del genere?

Ci sono alcuni eventi come `touchmove` su dispositivi *mobile* (quando l'utente sposta le dita lungo lo schermo), che di default causerebbero lo scrolling, ma ciò può essere evitato con `preventDefault()` sul gestore.

E così quando il browser rileva questo evento, prima elabora tutti i vari gestori, ed infine, solo nel caso in cui `preventDefault` non sia stato chiamato da nessuno dei gestori, procede con lo scrolling. Questa cosa potrebbe causare dei ritardi non voluti o dei "jitters" a video sulla UI.

L'opzione `passive: true` comunica al browser che il gestore non annullerà lo scrolling e il browser scrollerà la pagina immediatamente fornendo un'esperienza con la massima fluidità, con gli eventi che verranno gestiti all'occorrenza.

Per alcuni browser (Firefox, Chrome), `passive` è impostato a `true` di default per gli eventi `touchstart` e `touchmove`.

## event.defaultPrevented

La proprietà `event.defaultPrevented` sarà `true` se l'azione predefinita sarà stata prevenuta, altrimenti sarà `false`.

> C'è un relativo caso d'uso interessante.

Reviewer: ```suggestion
Relativamente a questo, c'è caso d'uso interessante.
```

Author: mi prendo l'appunto e correggo anche qui.

```html autorun height=50 no-beautify run
<button>Il click sul tasto destro mostra il menù contestuale</button>

<button *!*oncontextmenu="alert('Mostra il nostro menu'); return false"*/!*>
  Il click sul tasto destro mostra il nostro menù contestuale
</button>
```

Adesso, in aggiunta a questo menù, ci piacerebbe implementare un menù contestuale sul *documento*.

Al click sul tasto destro, dovrebbe comparire il relativo menù contestuale.

```html autorun height=80 no-beautify run
<p>Click sul tasto destro per il menù contestuale del documento</p>
<button id="elem">Clicca qui col tasto destro per il menu contestuale del pulsante</button>

<script>
  elem.oncontextmenu = function(event) {
    event.preventDefault();
    alert("Menu contestuale del pulsante");
  };

  document.oncontextmenu = function(event) {
    event.preventDefault();
    alert("Menu contestuale del documento");
  };
</script>
```

Il problema è che così facendo, cliccando su `elem`, otterremmo due menù: quello del pulsante (l'evento va risalendo per via del bubbling) e quello del documento.

> Come possiamo evitarlo? Una delle soluzioni è fare questo ragionamento: "Quando gestiamo il click sul tasto destro nel gestore del pulsante, stoppiamo il bubbling" e usiamo `event.stopPropagation()`:

Reviewer 2: ```suggestion
Come possiamo evitarlo? Una delle soluzioni è fare questo ragionamento: "Quando gestiamo il click sul tasto destro nel gestore del pulsante, interrompiamo il bubbling" e usiamo `event.stopPropagation()`:
```

```html autorun height=80 no-beautify run
<p>Click sul tasto destro per il documento</p>
<button id="elem">Click sul tasto destro per il menù del pulsante (sistemato con event.stopPropagation)</button>

<script>
  elem.oncontextmenu = function(event) {
    event.preventDefault();
*!*
    event.stopPropagation();
*/!*
    alert("Menu contestuale del pulsante");
  };

  document.oncontextmenu = function(event) {
    event.preventDefault();
    alert("Menu contestuale del documento");
  };
</script>
```

A questo punto il menù del pulsante funzionerà come previsto. Ma il prezzo sarà alto, perché a quel punto negheremo per sempre l'accesso alle informazioni relative ai click sul tasto destro, a qualunque altro codice esterno, inclusi contatori che raccologono statistiche e così via. Ed è una cosa poco saggia.

> Una soluzione alternativa potrebbe essere quella di controllare nel gestore del `document` se l'azione predefinita sia stata prevenuta. Se così fosse, significherebbe che l'evento è stato gestito, e quindi non sarà necessario gestirlo a sua volta.

Reviewer 2: ```suggestion
Una soluzione alternativa potrebbe essere quella di controllare nel gestore del `document` se l'azione predefinita sia stata prevenuta. Se così fosse, significherebbe che l'evento è stato gestito, e quindi non sarà necessario gestirlo nuovamente.
```

Author: /done

```

```smart header="Architettura dei menù contestuali annidati"
Ci sono pure dei modi alternativi per implementare i menù contestuali annidati. Uno di questi è quello di avere un singolo oggetto globale con un solo gestore per `document.oncontextmenu`, e metodi che ci permettono di gestire altri gestori al suo interno.

L'oggetto catturerà ogni click sul tasto destro, controllando tra i suoi gestori ed eseguire quello appropriato.

Ma in questo caso ogni pezzo di codice che vuole implementare un menù contestuale, dovrebbe conoscere l'esistenza di questo oggetto e del suo supporto, invece di avere il proprio gestore per `contextmenu`.
```

## Riepilogo

Ci sono tante azioni predefinite del borwser:

- `mousedown` -- comincia una selezione (spostare il mouse per continuare a selezionare).
- `click` su `<input type="checkbox">` -- check/uncheck sull'`input`.
- `submit` -- cliccando su `<input type="submit">` o premendo su `key:Enter` dentro un campo del form, scatena questo evento, ed il browser invia il form subito dopo.
- `keydown` -- premendo un tasto può portare ad aggiungere un carattere dentro un campo, o altre azioni.

> - `contextmenu` -- viene scatenato al click sul tasto destro, e l'azione che ne deriva è quella di mostarre il menù contestuale del browser.

Author: /done

> - ...e ce ne sono altri...

Reviewer 2: ```suggestion
- ...e molti altri...
```

Per prevenire un'azione predefinita -- si possono usare `event.preventDefault()` oppure  `return false`. Il secondo metodo è valido solo con gestori assegnati con `on<event>`.

L'opzione `passive: true` di  `addEventListener` comunica al browser che l'azione non sarà prevenuta. La sua utilità si palesa per eventi su dispositivi mobiles, come ad esempio `touchstart` e `touchmove`, per comunicare al browser che non deve attendere l'esecuzione di tutti i gestori prima di effettuare lo scrolling.

Se l'azione predefinita è stata prevenuta, il valore di `event.defaultPrevented` diventa `true`, altrimenti è `false`.

```warn header="Aderire alla semantica, non farne abuso"
Tecnicamente, prevenendo le azioni predefinite del browser e aggiungendo JavaScript, possiamo personalizzare il comportamento di qualunque elemento. Per esempio, possiamo fare in modo che un link `<a>` si comporti come un  pulsante, e un pulsante `<button>` come un link (redirezione su un altro URL o cose del genere).

Generalmente però, dovremmo mantenere il significato semantico degli elementi HTML. Per esempio, `<a>` dovrebbe comportare una navigazione, e non essere un pulsante.

> Oltre ad essere una "solo una cosa buona", ciò rende l'HTML migliore in termini di accessibilità.

Reviewer 2: Qui purtroppo la traduzione letterale non funziona troppo, crea solo più confusione

```suggestion
Non è "solamente una cosa buona", ciò rende l'HTML migliore in termini di accessibilità.
```

Author: /done

> Inoltre se consideriamo l'esempio con `<a>`, notiamo bene che: un browser ci permette di default di aprire questi links in una nuova finestra (cliccando sul tasto destro e con altri mezzi). E alla gente questo piace. Ma se invece creiamo un pulsante, che si comporta come se fosse un link usando JavaScript, e che appaia come se fosse un link con l'ausilio dei CSS, le funzionalità del browser, che specificatamente dedicate agli elementi `<a>`, non funzioneranno per il pulsante.

Reviewer 2: ```suggestion
Inoltre se consideriamo l'esempio con `<a>`, notiamo bene che: un browser ci permette di default di aprire questi links in una nuova finestra (cliccando sul tasto destro e con altri mezzi). E agli utenti questo piace. Ma se invece creiamo un pulsante, che si comporta come se fosse un link usando JavaScript, e che appaia come se fosse un link con l'ausilio dei CSS, le funzionalità del browser, che specificatamente dedicate agli elementi `<a>`, non funzioneranno per il pulsante.
```

Author: /done
